### DISCUSSION TITLE
게임 결과 정책

### DISCUSSION THREAD

**Discussion by tnwlsgk1002** [boostcampwm-2022/android11-almost-there#70] [Ideas]
- 게임 결과 화면에 어떤 부분을 포함시킬까?
- 종합 순위를 어떻게 정할까?
- 내기와 같은 요소를 낼 수 없을까?

**Comment by soulplay95** [comment]
## 종합 점수 정책

### 앱의 정체성

> 약속 시간에 지각하는 사람을 벌하는 컨셉

=> 제일 늦게 온 사람 또는 종합 점수에 따라 특정 비율로 나눠서 회비(술 값)을 더 부여한다.

### 점수에 반드시 포함되어야 할 요소

#### HP

- 우리의 자기장 정책 상 HP가 깎였다는 것은 약속 시간에 맞춰 도착할 확률이 매우 적다는 것을 의미한다.
    - HP가 깎였지만 제 시간에 도착하는 경우: 중간에 대중교통보다 빠른 수단(자동차, 비행기 등)을 이용해 무리해서 왔다.
 
#### 남은 직선 거리

- "제 시간에 도착 못함 == HP가 0" => 이것을 정책상 보장할 수 있는가?
- 그렇지 못할 때 약속 시간 시점에 각 유저의 현재 위치와 약속 장소 사이의 유클리드 거리를 구해서 점수에 포함시킨다.
- 가중치는 적절하게 정하기
- 약속 장소는 A 이지만, 미리 모인 사람끼리 B에서 모여 밥먹는 경우? => 도착 시 shake하여 도착했음을 알리는 기능 추가 => 이후 현재 위치 안받아옴(자기장 그릴 때 제외시킴)

### 도착 순서는 유의미한가?

- 약속 시간 안에만 도착하면 누가 먼저 왔는지는 점수에 영향이 없다.
    => 회비 덜내려고 무리하게 빨리 오려다가 사고날 위험 up

**Reply by tnwlsgk1002** [reply]
HP랑 남은 직선 거리로 계산하자는 거 동의합니다!
그리고 도착 시 shake하여 알리는 거 재밌을 것 같아요! 다만 이게 앱을 키고 있는 상태에서 해야 할 것 같고 공부가 필요할 것 같아서 우선순위는 뒤일 것 같다는 생각이 듭니다!

**Reply by fbghgus123** [reply]
처음 구상했던 아이디어들과 유사해서 좋아 보입니다.
도착지에서 Shake 해서 알리는 것도 좋아보여요 ! 👍

**Reply by tnwlsgk1002** [reply]
제시간에 도착하지 못했는데 HP가 100인 경우는
**약속 끝 시점에 자기장 원이 점이 된다면** 없을 것 같습니다.
HP가 깍이는 주기도 너무 길어지지 않게 조절해야할 것 같아요 
(10초 간격이면 좋겠지만, 너무 많이 잡아먹는다고 하면 실험해서 1분 이내로 조절하면 되지 않을까요?)

다만 다음과 같은 예외사항이 있을 것 같습니다.
>  **내 휴대폰에서 네트워크가 연결 되어있지 않으면 다른 사람 폰 에서는 HP가 깎이지 않는다.**

이 경우에는 중간에 연결된다면 HP가 연결되지만 처음을 제외하고 끝까지 네트워크를 끈다면에서 말할 수 있을 것 같습니다.

HP를 업데이트할 때 시간도 같이 포함을 시키기로 했으니
마지막 순간에 HP 업데이트 하고, **도착하지 않은 사람 중 업데이트가 안된 사람은 꼴찌 or 배제하면 좋지 않을까** 싶어요!

**Comment by tnwlsgk1002** [comment]
- 게임 결과 정책
  - 게임 만들 때 모드를 선택하며, **게임 상세 결과는 없다**
    - 상세 결과 화면 대신 종료 화면에서 캐릭터를 누르면 해당 캐릭터의 HP/도착 시간을 띄워준다.
- 만들 때 선택 : HP 우선 모드, 도착 시간 우선 모드
  - HP 우선 모드 : HP 순으로만 종합 순위 보기
  - 도착 시간 우선 모드 : 도착 시간 순으로만 종합 순위 보기
  - 종합 모드 : 도착 시간 / HP를 7:3으로 점수를 매김
- 결과 화면에서 선택 : 봐줄게 / 몰아주기 모드
  - 봐주기
    - 각각 순위를 곱해서 가져간다.
    - ex. 1등이 1배 2등이 2배 3등이 3배 => 1/6, 2/6, 3/6
  - 몰아주기
    - 꼴지한테 모두 몰아줌
    - 다이얼로그로 술값을 입력하면 계산하는 다이얼로그를 띄워준다.

**Comment by swjw13** [comment]
# 경우의수
## 시간 이전에 도착을 한 경우
- 이 경우에는 아무 penalty가 없어야 한다고 생각합니다.
## 시간이 다됐을 때도 도착을 못 한 경우
### HP 가 많이 닳지 않았다
- 시간 안에 도착을 하려고 노력했지만 실패 한 경우
- 도착지에 가까이 사는 경우
### HP가 많이 닳았다
- 적정 시간보다 늦게 출발 한 경우

<-- 이후 수정 예정 -->

**Comment by tnwlsgk1002** [comment]
# 정리
## 다수 의견
- 도착 시간은 배제하자
    - 시간 이전에 도착했으니 패널티를 부과하는 것은 아닌 것 같다.
- 도착한 사람은 무조건 1등이다. 
- 도착한 이후로는 업데이트를 하지 않는다.
    - 이 때 Shake로 하자는 의견
- 해당 비율에 따라 술값 내기
	- 1등이 1배, 2등이 2배, 3등이 3배... 식으로 부여한다.

## 선택 의견
> 도착하지 못한 사람끼리 HP만 가지고 결정하자 VS 남은 거리를 포함하자.
- HP만 가지고 결정하자는 의견
	- 우선 비율에 따라 정하지 않아도 되니 편리함
	- 남은 거리를 유클리드 거리로 정할 수 밖에 없으니 약간 애매한 부분도 있을 것 같다.
	- HP 자체가 자기장으로 결정되니까 약속 시간에 늦은 사람 중 **누가 늑장 부리냐**에 집중할 수 있을 것 같음
- 남은 거리를 포함하자는 의견
	- 과연 제시간에 도착하지 못할 때 HP == 100인 상황이 없을까?
		- 자기장 정책에 따라 다를 것 같다.
		- 그중 네트워크 이슈로 인한 문제는 HP 업데이트 시간으로 해결할 수 있다.

</br>

이렇게 나눠져서 투표를 진행하겠습니다!
HP는 ♥️ 남은 거리는 🚀으로 이모지를 남겨주시면 됩니다.

**Reply by tnwlsgk1002** [reply]
_**🎉HP만 가지고 랭킹을 매기기로 결정했습니다!🎉**_
![image](https://user-images.githubusercontent.com/76458724/204100977-fb25bff1-be1f-4eed-9fb3-bf3e802ecaee.png)
